### PR TITLE
C plus plus interface fix

### PR DESCRIPTION
**Pull request by zvookin**
Originally just a small fix for the code confusion reported in:
    https://github.com/halide/Halide/issues/1352
but went ahead and made all the C++ calling convention support work with the C++ backend. (Which is inaccurately called a C backend.) Added a test in the c_backend app.

### CODE REVIEW

**Comment by zvookin** [comment]
The C++ output should also be a bit cleaner if there are many externs with namespaces.

File: src/CodeGen_C.cpp [hunk 1]
```diff
@@ -161,7 +161,7 @@ const string globals =
     "}\n";
 }
 
-CodeGen_C::CodeGen_C(ostream &s, OutputKind output_kind, const std::string &guard) : IRPrinter(s), id("$$ BAD ID $$"), output_kind(output_kind) {
+  CodeGen_C::CodeGen_C(ostream &s, OutputKind output_kind, const std::string &guard) : IRPrinter(s), id("$$ BAD ID $$"), output_kind(output_kind), extern_c_open(false) {
```
Comment (steven-johnson) [inline src/CodeGen_C.cpp]: nit: did you mean to insert leading space?

File: src/CodeGen_C.cpp [hunk 2]
```diff
@@ -418,9 +422,27 @@ class ExternCallPrototypes : public IRGraphVisitor {
...
+
+    void emit_c_declarations(ostream &stream) {
+        for (const auto call : c_externs) {
+            emit_function_decl(stream, call.second, call.first);
+        }
+        stream << "\n";
+    }
+
+    void emit_c_plus_plus_declarations(ostream &stream) {
+        for (const auto ns_or_call : c_plus_plus_externs) {
```
Comment (steven-johnson) [inline src/CodeGen_C.cpp]: should this be "const auto &" (here and elsewhere)?

File: apps/c_backend/pipeline_cpp.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+// Make sure extern "C" works
+HalideExtern_2(int, an_extern_c_func, int, float);
+
+int main(int argc, char **argv) {
+    Func f;
+    ImageParam input(UInt(16), 2);
+    Var x, y;
+
+    Expr add_all_the_things = cast<int>(0);
```
Comment (steven-johnson) [inline apps/c_backend/pipeline_cpp.cpp]: This name left me hoping for a Hyperbole And A Half shout-out. Left disappointed.

**Pull request merged**